src/Nonlinear/types.jl

```diff
@@ -0,0 +1,237 @@
+    NODE_MOI_VARIABLE,
+    # Index of the internal, consecutive, and ordered `MOI.VariableIndex`.
+    NODE_VARIABLE,
+    # Index is into the list of constants
+    NODE_VALUE,
+    # Index is into the list of parameters
+    NODE_PARAMETER,
+    # Index is into the list of subexpressions
+    NODE_SUBEXPRESSION,
+)
+
+"""
+    struct Node
+        type::NodeType
+        index::Int
+        parent::Int
+    end
+
+A single node in a nonlinear expression tree. Used by
+[`NonlinearExpression`](@ref).
+
+See the MathOptInterface documentation for information on how the nodes and
+values form an expression tree.
+"""
+struct Node
+    type::NodeType
+    index::Int
+    parent::Int
+end
+
+"""
+    struct NonlinearExpression
+        nodes::Vector{Node}
+        values::Vector{Float64}
+    end
+
+The core type that represents a nonlinear expression. See the MathOptInterface
+documentation for information on how the nodes and values form an expression
+tree.
+"""
+struct NonlinearExpression
+    nodes::Vector{Node}
+    values::Vector{Float64}
+    NonlinearExpression() = new(Node[], Float64[])
+end
+
+function Base.:(==)(x::NonlinearExpression, y::NonlinearExpression)
+    return x.nodes == y.nodes && x.values == y.values
+end
+
+"""
+    struct NonlinearConstraint
+        expression::NonlinearExpression
+        set::Union{
+            MOI.LessThan{Float64},
+            MOI.GreaterThan{Float64},
+            MOI.EqualTo{Float64},
+            MOI.Interval{Float64},
+        }
+    end
+
+A type to hold information relating to the nonlinear constraint `f(x) in S`,
+where `f(x)` is defined by `.expression`, and `S` is `.set`.
+"""
+struct NonlinearConstraint
+    expression::NonlinearExpression
+    set::Union{
+        MOI.LessThan{Float64},
+        MOI.GreaterThan{Float64},
+        MOI.EqualTo{Float64},
+        MOI.Interval{Float64},
+    }
+end
+
+"""
+    ParameterIndex
+
+An index to a nonlinear parameter that is returned by [`add_parameter`](@ref).
+Given `data::NonlinearData` and `p::ParameterIndex`, use `data[p]` to retrieve
+the current value of the parameter and `data[p] = value` to set a new value.
+"""
+struct ParameterIndex
+    value::Int
+end
+
+"""
+    ExpressionIndex
+
+An index to a nonlinear expression that is returned by [`add_expression`](@ref).
+
+Given `data::NonlinearData` and `ex::ExpressionIndex`, use `data[ex]` to
+retrieve the corresponding [`NonlinearExpression`](@ref).
+"""
+struct ExpressionIndex
+    value::Int
+end
+
+"""
+    ConstraintIndex
+
+An index to a nonlinear constraint that is returned by [`add_constraint`](@ref).
+
+Given `data::NonlinearData` and `c::ConstraintIndex`, use `data[c]` to
+retrieve the corresponding [`NonlinearConstraint`](@ref).
+"""
+struct ConstraintIndex
+    value::Int
+end
+
+"""
+    NonlinearData()
+
+Construct a new [`MOI.AbstractNLPEvaluator`](@ref) object.
+"""
+mutable struct NonlinearData <: MOI.AbstractNLPEvaluator
+    inner::Union{Nothing,MOI.AbstractNLPEvaluator}
```

Comment:
Would it make sense for `NonlinearData` to be simply data (as its name suggests) and not an `AbstractNLPEvaluator`? The data could be an inner field of the evaluator. I'm aware it gets a bit complicated because of MOI API and how evaluators are selected.

Reply:
I made it an `AbstractNLPEvaluator` so that it can be directly used for the expression graph stuff, and so we can unify our timers of the different callbacks. Otherwise we'd just need an additional layer of indirection.

Reply:
Eventually `NonlinearData` will be first class and accessible to extensions, right? If so, shouldn't this struct be clean and not have cruft like internal timers? How will `NonlinearData` interact with `NLPBlockData`? There's already a lot of layers (NLPBlockData -> NonlinearData -> NonlinearData.inner).

Reply:
Okay, I'll take a pass at re-organizing this.